### DISCUSSION TITLE
v0.6-dev - binary size increase

### DISCUSSION THREAD

**Discussion by michaelherger** [librespot-org/librespot#1415] [General]
I see that the binaries for 0.6-dev have increased in size considerably compared to the previous 0.6. A release build on macOS/arm64 would now be up to 7.2MB from 4.3MB. Is this to be expected? Or did I miss something?

**Comment by photovoltex** [comment]
Could you include your compile command? Or is it without any adjustments?

**Reply by michaelherger** [reply]
It's admittedly my own fork. I'm currently going through dependencies, to see whether after merging 0.6-dev I'm importing more than I did before. Haven't even pushed it to Github yet... (https://github.com/michaelherger/librespot/tree/spotty-dev)

**Reply by photovoltex** [reply]
It might be because we include new protobuf definitions for the dealer. Or related to some dependency changes. I'm currently unable to check, but you could try to exclude the dealer commit and check if that makes the difference.

**Reply by michaelherger** [reply]
FWIW: stripped un-touched `dev` branch is about 15MB vs. 12MB before dealer. So some growth is there in that work. It's possible that I did exclude some stuff before, which now has been re-introduced by all the refactoring. Thanks a ton for your work anyway!

**Reply by photovoltex** [reply]
Seems to match up to your reported 3MB size increase. I will take a look into it later.

Wouldn't be surprised if I included too many proto definitions by accident. But I'm kinda surprised that these changes made such a big impact on the size of the binary. Might also just be rust... but will see

**Reply by photovoltex** [reply]
I'm currently investigating the topic a bit on the side and the following diff already makes a ~1.3MB diff (stripped release build on linux). I will continue investigating but maybe someone already see's where the problem lies, because I don't see any direct issue with these changes that should cause a 1.3MB diff.
https://github.com/librespot-org/librespot/compare/67d31959f5125d3fe5fab5e6e95436a01bb3cef6...2ecb8f06a781457710ae678bf4988cf4856047de

**Reply by photovoltex** [reply]
Okay it seems this here caused a 1MB file size increase. https://github.com/librespot-org/librespot/commit/dfdfbd57e27ef7af0422e65006a20c73d6cad16a

In general it seems that file addition increase the size of the binary

**Reply by roderickvd** [reply]
Just for kicks, what happens when you compile in release profile with full link time optimizations?

In `Cargo.toml` or `~/.cargo/config.toml`:
```toml
[profile.release]
lto = true
codegen-units = 1
```

Or on the command line: `RUSTFLAGS="-Clto Ccodegen-units=1 cargo build --release`

**Reply by photovoltex** [reply]
Ah nice hint. After I just saw that adding files seem to increase the binary size a lot, I wanted to continue investigating these four commits by improving the optimization flags and see if that maybe improves the size. Will of course try these flags next :)

I also just remembered that there was this repository https://github.com/johnthagen/min-sized-rust

**Reply by michaelherger** [reply]
I’ve always been using `lto` in my fork. Though I thought it was not recommended in general because it removes a lot of information you might want to have to investigate runtime issues. But I’m using anyway. The size increase I noticed with this parameter in place too.

**Reply by photovoltex** [reply]
Hmm oki, good to already know what I might see. So far I can see noticeable file size increases (around 100KB-200KB) when adding files. I suppose the 1MB change is when doing more then just adding a file. It would be a bit weird to have a language that only works best with single files.

**Reply by photovoltex** [reply]
@michaelherger I just did a quick search for `lto` and I didn't found anything that would speak against the use of the flag? Do you have any reference to that recommendation that you mentioned? Personally I don't see any harm in decreasing the size of the binary. And as long as the logger still works (and we have logging at the correct positions) it should be fine in my opinion.

**Reply by roderickvd** [reply]
You can strip debug info but I don't think that even happens by default in release profile:

https://doc.rust-lang.org/cargo/reference/profiles.html#strip
https://doc.rust-lang.org/cargo/reference/profiles.html#release

**Reply by photovoltex** [reply]
And here are my final finds on the topic https://docs.google.com/spreadsheets/d/1eSYGozN89wzfH2lV9l6tfL6nHr-iprYG24ORtQa50eM.

The binary seem to have increased in size ~0.9MB when optimized with

```toml
[profile.release]
strip = true
lto = true
codegen-units = 1
opt-level = 'z'
panic = 'abort'
```

I think this might be _alright_, as we added `flate2` and `protobuf_json_mapping` and the changes where still quite big. I might also be wrong as I never optimized rust binaries before.

The last thing we could still investigate is on how much the dependencies and the code itself changed the size. If I find some time, I will take a look at it again as the topic seems quite interesting hehe.

**Reply by roderickvd** [reply]
When using `tokio` I think it's best to keep `panic` as `unwind`, to let it `catch_unwind` in various places to clean up.

**Reply by michaelherger** [reply]
> @michaelherger I just did a quick search for `lto` and I didn't found anything that would speak against the use of the flag? Do you have any reference to that recommendation that you mentioned? Personally I don't see any harm in decreasing the size of the binary. And as long as the logger still works (and we have logging at the correct positions) it should be fine in my opinion.

Maybe I confused it with stripping the binary. In any case I haven't seen much of an issue in the several years I've been using it.

That said, I believe this is not the point about the topic I've raised. What I was seeing is that the file size increased considerably, all given the same parameters and build options. 

Hopefully I'll find some time over the Holidays to play with https://github.com/librespot-org/librespot/compare/67d31959f5125d3fe5fab5e6e95436a01bb3cef6...2ecb8f06a781457710ae678bf4988cf4856047de. Is this part of a PR I could check out?

**Reply by photovoltex** [reply]
The slice contains the first four commits of the dealer rework. I still have local copies of the branch so I can easily checkout the revisions. I will push the branch later so that you can investigate yourself (https://github.com/photovoltex/librespot/tree/merged/integrate-dealer). I also think the branch is not complete, but as the first four commits already make such a big difference it's maybe already enough to find the root cause.
